File: Back/NodeJS/modules/bets/src/infrastructure/integrationEvents/IntegrationEventFactory.ts
import { Outbox } from "../../../../shared/infrastructure/outbox/Outbox";
import { IIntegrationEvent } from '../../../../shared/infrastructure/integrationEvents/IIntegrationEvent';
import { BetCompleted } from "../../domain/bets/events/BetCompleted";
import { BetCreated } from "../../domain/bets/events/BetCreated";
import { BetCreatedIntegrationEvent } from './BetCreatedIntegrationEvent';
import { BetCompletedIntegrationEvent } from './BetCompletedIntegrationEvent';

export class IntegrationEventFactory {
    Create(outbox: Outbox): IIntegrationEvent {
        switch(outbox.Type) {
            case BetCompleted.name:
                const betCompleted = JSON.parse(outbox.Data) as BetCompleted
                return new BetCompletedIntegrationEvent(betCompleted.BetId.Value, betCompleted.IsSuccessful)
            case BetCreated.name:
                const betCreated = JSON.parse(outbox.Data) as BetCreated
                return new BetCreatedIntegrationEvent(betCreated.BetId.Value)
            default:
                return undefined;
        }
    }

}

File: Back/.Net/Modules/Bets/BetFriends.Bets.Infrastructure/IntegrationEvents/IntegrationEventFactory.cs
using BetFriends.Bets.Domain.Events;
using BetFriends.Shared.Infrastructure.IntegrationsEvents.Abstractions;
using BetFriends.Shared.Infrastructure.Outboxes;
using System.Text.Json;

namespace BetFriends.Bets.Infrastructure.IntegrationEvents;

public class IntegrationEventFactory
{
    public IIntegrationEvent Create(Outbox outbox)
    {
        switch (outbox.Type)
        {
            case nameof(BetCreated):
                var betCreated = JsonSerializer.Deserialize<BetCreated>(outbox.Data)!;
                return new BetCreatedIntegrationEvent(betCreated.BetId.Value, betCreated.OwnerId.Value, betCreated.Coins);
            case nameof(BetCompleted):
                var betCompleted = JsonSerializer.Deserialize<BetCompleted>(outbox.Data)!;
                return new BetCompletedIntegrationEvent(betCompleted.BetId.Value, betCompleted.IsSuccessful);
            default:
                return null!;
        }
    }
}
